### PR TITLE
Vlib impl private modules

### PR DESCRIPTION
**Pull request by rgrinberg**
Two test cases for a couple of other important situations with virtual libraries:

* Private modules in implementations and virtual libraries should never overlap (right now this doesn't work)

* Implementation that provide a lib interface module should be alerted that it isn't possible. This isn't obvious when virtual libraries' LIM is the alias module. It's already not possible, but we should add a proper error message explaining the situation.

### CODE REVIEW

File: test/blackbox-tests/test-cases/variants/run.t [hunk 1]
```diff
@@ -202,3 +202,25 @@ Implementations may refer to virtual library's modules
           test alias default
   bar from vlib
   Foo.run implemented
+
+Implementations may have private modules that have overlapping names with the
+implementation lib
```
Comment (diml) [inline test/blackbox-tests/test-cases/variants/run.t]: Shouldn't this be:
```suggestion
virtual lib
```

Comment (emillon) [reply test/blackbox-tests/test-cases/variants/run.t]: I'm loving this github feature already.

Comment (diml) [reply test/blackbox-tests/test-cases/variants/run.t]: Yh, I was pleasantly surprised by it this morning

**Comment by rgrinberg** [comment]
The apply suggestion thing doesn't include the sign-off. So I suppose it's kind of useless for us :(

**Comment by rgrinberg** [comment]
@diml i applied your suggestion manually

**Comment by rgrinberg** [comment]
Okay, the last few commits make this PR ready for review. All problems related to private modules should be solved and the code should be easier to understand as well.

One minor blemish is the code duplication of code introduced by `Ocamldep.Dep_graphs.Multi`. The old top closure could be implemented on top of this more general approach, but it's slower and harder to read so a little bit of duplication seems to be the lesser evil.

File: src/ocamldep.ml [hunk 2]
```diff
@@ -42,6 +45,32 @@ module Dep_graph = struct
...
+  module Multi = struct
+    let top_closed_multi (ts : t list) modules =
+      List.concat_map ts ~f:(fun t ->
+        Module.Name.Map.to_list t.per_module
+        |> List.map ~f:(fun (_name, (unit, deps)) ->
+          deps >>^ fun deps -> (unit, deps)))
+      |> Build.all >>^ fun per_module ->
+      let per_obj =
+        Module.Obj_map.of_list_reduce per_module ~f:List.rev_append in
+      match Module.Obj_map.top_closure per_obj modules with
```
Comment (diml) [inline src/ocamldep.ml]: We should be able to factorise this code with the one in `top_closed`

**Comment by diml** [comment]
This looks better indeed. We should be able to factorise at least the error reporting code

File: src/ocamldep.ml (hunk 2 above)
Comment (rgrinberg) [reply src/ocamldep.ml]: The errors are actually a bit different as one contains a directory. I've factorised the pretty printing of the cycle and the other function though.

**Pull request merged**